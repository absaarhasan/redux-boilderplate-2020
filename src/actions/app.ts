import * as actions from 'Actions/constants';
import { Notification } from '../utils/index';
import {
  IsLoadingAction,
  IsBusyAction,
  SetNotificationAction,
  ClearNotificationAction,
} from './types';

export const isLoadingAction = (value: boolean): IsLoadingAction => ({
  type: actions.APP_IS_LOADING,
  isLoading: value,
});

export const isBusyAction = (value: boolean): IsBusyAction => ({
  type: actions.APP_IS_BUSY,
  isBusy: value,
});

export const setNotificationAction = (data: Notification): SetNotificationAction => ({
  type: actions.NOTIFICATION_SET,
  data,
});

export const clearNotificationAction = (): ClearNotificationAction => ({
  type: actions.NOTIFICATION_CLEAR,
});
