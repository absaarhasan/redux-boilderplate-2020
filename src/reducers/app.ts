import { AppActionTypes } from 'Actions/types';
import * as actions from 'Actions/constants';
import { AppState } from './types';

const initialState: AppState = {
  isLoading: false,
  isBusy: false,
};

const app = (state = initialState, action: AppActionTypes): AppState => {
  switch (action.type) {
    case actions.APP_IS_LOADING:
      return {
        ...state,
        isLoading: action.isLoading,
      };
    case actions.APP_IS_BUSY:
      return {
        ...state,
        isBusy: action.isBusy,
      };
    default:
      return state;
  }
};

export default app;
