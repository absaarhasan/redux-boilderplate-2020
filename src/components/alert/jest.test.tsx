import React from 'react';
import {
  cleanup, screen, fireEvent,
} from '@testing-library/react';
import * as ReactRedux from 'react-redux';
import * as actions from 'Actions/app';
import { render } from '../../../jest.setup';
import '@testing-library/jest-dom';

import Alert from '.';


afterEach(cleanup);

describe('<Alert /> spec', () => {
  it('should render', () => {
    const initialState = {
      notifications: {
        active: true,
        variant: 'error',
        msg: 'This is an eror.',
      },
    };
    render(<Alert />, initialState);
    expect(screen.getByTestId('alert-container')).toMatchSnapshot();
  });
  it('should be hidden by default', () => {
    render(<Alert />);
    expect(screen.queryByTestId('alert-container')).toBeNull();
  });
  it('should display when triggered', () => {
    const initialState = {
      notifications: {
        active: true,
        variant: 'error',
        msg: 'This is an eror.',
      },
    };
    render(<Alert />, initialState);
    expect(screen.getByTestId('alert-container')).toBeDefined();
  });
  it('should display the alert message', () => {
    const initialState = {
      notifications: {
        active: true,
        variant: 'error',
        msg: 'This is an eror.',
      },
    };
    render(<Alert />, initialState);
    expect(screen.getByTestId('msg').innerHTML).toBe('This is an eror.');
  });
  it('the close button shoud dispatch the correct action', () => {
    const initialState = {
      notifications: {
        active: true,
        variant: 'error',
        msg: 'This is an eror.',
      },
    };
    const useDispatch = jest.spyOn(ReactRedux, 'useDispatch');
    const actionCreator = jest.spyOn(actions, 'clearNotificationAction');
    render(<Alert />, initialState);
    fireEvent.click(screen.getByTestId('close-btn'));
    expect(useDispatch).toHaveBeenCalled();
    expect(actionCreator).toHaveBeenCalled();
  });
  it('the close button should update the global state', () => {
    const initialState = {
      notifications: {
        active: true,
        variant: 'error',
        msg: 'This is an eror.',
      },
    };
    const expectedState = { active: false, variant: null, msg: '' };
    const { store } = render(<Alert />, initialState);
    fireEvent.click(screen.getByTestId('close-btn'));
    expect(store.getState().notifications).toEqual(expectedState);
  });
});
