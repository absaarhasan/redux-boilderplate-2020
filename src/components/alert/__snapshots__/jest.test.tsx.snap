// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> spec should render 1`] = `
.emotion-0 {
  width: 100%;
  min-height: 100px;
  position: relative;
  text-align: center;
}

.emotion-0 h2 {
  margin: 20px 0 0 0;
  color: #ffffff;
  display: inline-block;
}

.emotion-0 .close-btn {
  position: absolute;
  right: 0;
  height: 20px;
  width: 20px;
}

.emotion-0.success {
  background-color: #488127;
}

.emotion-0.warn {
  background-color: #f0a534;
}

.emotion-0.error {
  background-color: #df142f;
}

<div
  class="error emotion-0"
  data-testid="alert-container"
>
  <button
    class="close-btn"
    data-testid="close-btn"
    type="button"
  >
    X
  </button>
  <h2
    data-testid="msg"
  >
    This is an eror.
  </h2>
</div>
`;
